### ISSUE TITLE
CompiledDataBinder: "Body of catch must have the same type as body of try" exception

### ISSUE THREAD

**Issue by ChairmanMawh** [MarkPflug/Sylvan#224]
Not too sure what to do with this one (not a big user of expressions)..

![image](https://github.com/MarkPflug/Sylvan/assets/118385137/85887d92-b7cd-4c3d-a0d9-cba7549e00ba)

This console app reproduces it with 1.3.5, 0.2.12, and .net 7

```
        static bool HandleError(DataValidationContext validationContext)
        {
            return true;
        }

        public class X{
            public string UniqueIdentifier { get; set; }
            public DateTime RecordDate { get; set; }
            public decimal? Data1 { get; set; }
        }

        public static void Main(string[] args)
        {
            var csvStr = @"Z, Y, ID, X, W, V, A1, A2
""A"", ""B"", ""C"", ""D"", ""E"", 01/01/2000, 12.345, 67.890"u8;
            var schema = "ID>UniqueIdentifier:string,V>RecordDate:DateTime{dd/MM/yyyy},A1>Data1:decimal?";

            var opts = new CsvDataReaderOptions { Schema = new CsvSchema(Sylvan.Data.Schema.Parse(schema)), HasHeaders = true };
            using var csv = CsvDataReader.Create(new StreamReader(new MemoryStream(csvStr.ToArray())), opts).ValidateSchema(HandleError);

            var binderOpts = new DataBinderOptions() { BindingMode = DataBindingMode.Any };
            var l = csv.GetRecords<X>(binderOpts).ToList(); //boom
    }
```

**Comment by ChairmanMawh** [comment]
ps; as noted I'd no idea what to do with it, but implementing BingAI's recommendation did at least stop it from throwing.. No idea if it's right or not

![image](https://github.com/MarkPflug/Sylvan/assets/118385137/84094e5e-951e-42c7-b0d1-a6816eebf15d)


```
var catchExpr = Expression.Catch(exParam, Expression.Block(Expression.Throw(Expression.New(exCtor!, idxVar, exParam)), Expression.Default(typeof(void))));
```

**Comment by ChairmanMawh** [comment]
Pumelling a couple of AIs eventualy got to the suggestion to change the code to:

```
var catchExpr = Expression.Catch(
			exParam,
			Expression.Block(
				Expression.Throw(Expression.New(exCtor!, idxVar, exParam)),
				Expression.NewArrayInit(typeof(object)) // Return an empty array of objects
			)
		); 
```

which seems to stop the exception throwing, but no idea if it's right..

**Comment by MarkPflug** [comment]
Sounds like you got it figured out?

The original issue is that there was *nothing* getting bound, because the headers had a space prefixed, which probably should have produced a binding error. Your code uses `BindingMode.Any`, which currently allows zero properties to be bound. Technically, it might make sense to allow a completely uninitialized `T` to be returned, since `Any` includes none, but I think in reality it would always mask a mistake of the user.

Would it have been more obvious if an `UnboundMemberException` had been thrown originally?

**Comment by ChairmanMawh** [comment]
Yeah, I've resolved to start off with bindingmode All in future, and then swap it to Any.. Even when I was debugging Sylvan's code today, I'm looking at a vis of the physicalColumns var and pulling my hair out wondering "why are there 54 entities here and every single one of them has no name at all.. infact, no anything at all - all the properties are the default values?" - it was so easy to miss the leading space in the data vs the schema

"Any" probably should have a special case for when nothing matches, and perhaps raise an error if a schema mentions a property that doesn't exist - while we can't reasonably control whether a column is or isn't in a CSV from one file to the next, the compiled code nominating `A>B` could raise an error if B doesn't exist as a proeprty in a compiled class.. MissingMemberException?

**Comment by ChairmanMawh** [comment]
Actually, thinking on it some more, I think my requirement to use Any is a consequence of not having available the mode that I really, actually want to use..

All notionally means "All columns in file related to All properties in model" but I have columns I don't want to use and for some files I have properties I cannot set (no data for them) but it doesn't matter for those props. I can't use AllColumns or AllProps either for the same reason; sometimes a file will have both unbound cols and result in unassigned props
.. and thus my only choice is to use Any, which doesn't check anything

Really, I'd like the schema, if I've used one, to be the driver for an All operation; if I've put columns in the schema that don't exist in the file, or i've mentioned props in the schema that don't exist in the class, then I should get an error so I can correct the code

In other words where we now have All/AllX driven by the file/class, have a different mode/setting where it's driven by the schema - do an All, but pretend that columns/props not mentioned in the schema don't exist

**Comment by MarkPflug** [comment]
The data binder can use the `DataMemberAttribute` to allow mapping names that don't match the property. That attribute has an `IsRequired` member, but I don't currently look at it. I think it would make sense to enforce the binder binds all `IsRequired` properties regardless of BindingMode.

**Comment by MarkPflug** [comment]
I've made two changes.
1) the binder *must* bind at least one property regardless of binding mode or an UnboundMemberException will be thrown.
2) the binder now honors the DataMemberAttribute.IsRequired property, such that any properties with that flag *must* be bound, or an UnboundMemberException will be thrown.

These changes are included in this pre-release package:
https://www.nuget.org/packages/Sylvan.Data/0.2.13-b0001

I'll release a 0.2.13 final release at some point in the near future.
